README.md
yavdr-addon-irmp
================

Driver and Skripts for USB-IRMP Receiver


debian/changelog
irmplircd (2.2.3-git2013112-17yavdr0~precise) precise; urgency=medium

  * fix two typos in irmplircd2uinput.conf (thanks to raiuns)

 -- Alexander Grothe <seahawk1986@gmx.de>  Tue, 21 Jan 2014 15:53:34 +0100

irmplircd (2.2.3-git2013112-16yavdr0~precise) precise; urgency=medium

  * add missing "$" to expand variable in irmplircd.upstart exec line 

 -- Alexander Grothe <seahawk1986@gmx.de>  Tue, 21 Jan 2014 10:59:46 +0100

irmplircd (2.2.3-git2013112-15yavdr0~precise) precise; urgency=medium

  * next try to fix argument quoting 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 22:33:33 +0100

irmplircd (2.2.3-git2013112-14yavdr0~precise) precise; urgency=medium

  * rearrange arguments for irmpexec 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 22:10:12 +0100

irmplircd (2.2.3-git2013112-13yavdr0~precise) precise; urgency=medium

  * change argument formatting for irmplircd 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 22:03:41 +0100

irmplircd (2.2.3-git2013112-12yavdr0~precise) precise; urgency=medium

  * fix environment variable definition (must not be done in pre-start!) 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 21:16:23 +0100

irmplircd (2.2.3-git2013112-11yavdr0~precise) precise; urgency=medium

  * update udev rule 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 18:37:30 +0100

irmplircd (2.2.3-git2013112-10yavdr0~precise) precise; urgency=medium

  * update udev rule 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 18:31:38 +0100

irmplircd (2.2.3-git2013112-9yavdr0~precise) precise; urgency=medium

  * update udev rule, cleanup upstart file 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 17:40:12 +0100

irmplircd (2.2.3-git2013112-8yavdr0~precise) precise; urgency=medium

  * use /etc/default/irmplircd as config file for irmplircd2uinput Upstart job 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 16:16:51 +0100

irmplircd (2.2.3-git2013112-7yavdr0~precise) precise; urgency=medium

  * update upstart file 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 15:50:51 +0100

irmplircd (2.2.3-git2013112-6yavdr0~preciseubuntu1) precise; urgency=low

  * update udev rule

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 15:13:55 +0100

irmplircd (2.2.3-git2013112-6yavdr0~precise) precise; urgency=medium

  * fix upstart file, so install succeedes 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 15:06:02 +0100

irmplircd (2.2.3-git2013112-5yavdr0~precise) precise; urgency=medium

  * add README.Debian 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 14:40:59 +0100

irmplircd (2.2.3-git2013112-4yavdr0~precise) precise; urgency=medium

  * reorganize upstart files 

 -- Alexander Grothe <seahawk1986@gmx.de>  Mon, 20 Jan 2014 13:28:37 +0100

irmplircd (2.2.3-git2013112-3yavdr0~precise) precise; urgency=medium

  * add manpage to package

 -- Alexander Grothe <seahawk1986@gmx.de>  Fri, 17 Jan 2014 13:12:54 +0100

irmplircd (2.2.3-git2013112-2yavdr0~precise) precise; urgency=medium

  * fix order of arguments in upstart script 

 -- Alexander Grothe <seahawk1986@gmx.de>  Fri, 17 Jan 2014 13:07:52 +0100

irmplircd (2.2.3-git2013112-1yavdr0~precise) precise; urgency=medium

  * change default lircd-socket to /var/run/lirc/irmplircd 

 -- Alexander Grothe <seahawk1986@gmx.de>  Fri, 17 Jan 2014 12:58:32 +0100

irmplircd (2.2.3-git2013112-0yavdr0~precise) precise; urgency=medium

  * rename to irmplircd 

 -- Alexander Grothe <seahawk1986@gmx.de>  Fri, 17 Jan 2014 12:28:31 +0100

easyvdr-usb-irmp (2.2.3-git2013112-3yavdr0~precise) precise; urgency=medium

  * increase version number, change upstart file 

 -- Alexander Grothe <seahawk1986@gmx.de>  Fri, 17 Jan 2014 12:20:17 +0100

easyvdr-usb-irmp (2.2.3-git2013112-0yavdr0~precise) precise; urgency=low

  * fix hashmap.h include 

 -- Alexander Grothe <seahawk1986@gmx.de>  Fri, 17 Jan 2014 12:11:33 +0100

easyvdr-usb-irmp (2.2.3-git2013112-2easyVDR0~trusty) trusty; urgency=high

  * integration in easyvdr

 -- easyVDR-Developer <easyvdr-dev@easy-vdr.de>  Thu, 02 Jan 2014 09:58:42 +0100

easyvdr-usb-irmp (2.2.3-git2013112-1easyVDR2~trusty) trusty; urgency=high

  * renamed package

 -- easyVDR-Developer <easyvdr-dev@easy-vdr.de>  Sun, 29 Dec 2013 17:49:55 +0100

usb-irmp (2.2.3-git2013112-1easyVDR1~trusty) trusty; urgency=high

  * changed symlink, depend

 -- easyVDR-Developer <easyvdr-dev@easy-vdr.de>  Mon, 09 Dec 2013 13:24:50 +0100

usb-irmp (2.2.3-git2013112-0easyVDR2~trusty) trusty; urgency=high

  * fix for missing map file

 -- easyVDR-Developer <easyvdr-dev@easy-vdr.de>  Sun, 08 Dec 2013 17:32:26 +0100

usb-irmp (2.2.3-0easyVDR10~trusty) trusty; urgency=high

  * Initial release

 -- easyVDR-Developer <easyvdr-dev@easy-vdr.de>  Wed, 20 Nov 2013 15:58:08 +0100


mapping.c
/*
    irmplircd -- zeroconf LIRC daemon that reads IRMP events from the USB IR Remote Receiver
	             http://www.mikrocontroller.net/articles/USB_IR_Remote_Receiver
    Copyright (C) 2011-2012  Dirk E. Wagner

	based on:
    inputlircd -- zeroconf LIRC daemon that reads from /dev/input/event devices
    Copyright (C) 2006  Guus Sliepen <guus@sliepen.eu.org>

    This program is free software; you can redistribute it and/or modify it
    under the terms of version 2 of the GNU General Public License as published
    by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
*/

 /* Standard headers */
#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sysexits.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <syslog.h>
#include <ctype.h>

#include "debug.h"
#include "c_hashmap/hashmap.h"
#include "mapping.h"

bool parse_translation_table(const char *path, map_t mymap) {
	FILE *table;
	char *line = NULL;
	size_t line_size = 0;
	char key[KEY_MAX_LENGTH];
	char value[KEY_MAX_LENGTH];
	int error = 0;
	int len;

	key[0] = value[0] = 0;	
	if(!path)
		return false;

	table = fopen(path, "r");
	if(!table) {
		fprintf(stderr, "Could not open translation table %s: %s\n", path, strerror(errno));
		return false;
	}

	while(getline(&line, &line_size, table) >= 0) {
		len = sscanf(line, "%99s %99s", key, value);
		if(len != 2) {
			syslog(LOG_ERR, "line ignored: %s\n", line);
			continue;
		}
		key[KEY_MAX_LENGTH-1] = '\0';
		value[KEY_MAX_LENGTH-1] = '\0';
		
		if(strlen(key) < 1 || strlen(value) < 1)
			continue;

		DBG ("parse_translation_table: key = %s, value = %s\n", key, value);
		
		map_entry_t *map_entry = malloc(sizeof(map_entry_t));
  		snprintf(map_entry->key, KEY_MAX_LENGTH, "%s", key);
		snprintf(map_entry->value, KEY_MAX_LENGTH, "%s", value);

		error = hashmap_put(mymap, map_entry->key, map_entry);			

		if(error) {
			fprintf(stderr, "hashmap_put failure: %d\n", error);
			fclose(table);
			free(line);
			return false;
		}
	}

	fclose(table);
	free(line);
	
	return true;
}


irmplircd.c
/*
    irmplircd -- zeroconf LIRC daemon that reads IRMP events from the USB IR Remote Receiver
	             http://www.mikrocontroller.net/articles/USB_IR_Remote_Receiver
    Copyright (C) 2011-2012  Dirk E. Wagner

	based on:
    inputlircd -- zeroconf LIRC daemon that reads from /dev/input/event devices
    Copyright (C) 2006  Guus Sliepen <guus@sliepen.eu.org>

    This program is free software; you can redistribute it and/or modify it
    under the terms of version 2 of the GNU General Public License as published
    by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
*/

#define _GNU_SOURCE
#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/time.h>
#include <sysexits.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <syslog.h>
#include <pwd.h>
#include <ctype.h>

/* Input subsystem interface */
#include <linux/input.h>

#include "debug.h"
#include "c_hashmap/hashmap.h"
#include "mapping.h"

typedef struct {
  uint8_t	dummy;		// additional useless byte 
  uint8_t	protocol;	// protocol, i.e. NEC_PROTOCOL
  uint16_t	address;	// address
  uint16_t	command;	// command
  uint8_t	flags;		// flags, e.g. repetition
} IRMP_DATA;

typedef struct evdev {
	char *name;
	int fd;
	struct evdev *next;
} evdev_t;

static evdev_t *evdevs = NULL;

typedef struct client {
	int fd;
	struct client *next;
} client_t;

static client_t *clients = NULL;

static int sockfd = -1;

static bool grab = false;
static char *device = "/var/run/lirc/lircd";

static long repeat_time = 0L;
static struct timeval previous_input;
static int repeat = 0;

static map_t mymap;

static void *xalloc(size_t size) {
	void *buf = malloc(size);
	if(!buf) {
		fprintf(stderr, "Could not allocate %zd bytes with malloc(): %s\n", size, strerror(errno));
		exit(EX_OSERR);
	}
	memset(buf, 0, size);
	return buf;
}

static void add_evdevs(int argc, char *argv[]) {
	int i;
	evdev_t *newdev;

	for(i = 0; i < argc; i++) {
		newdev = xalloc(sizeof *newdev);
		newdev->fd = open(argv[i], O_RDONLY);
		if(newdev->fd < 0) {
			free(newdev);
			fprintf(stderr, "Could not open %s: %s\n", argv[i], strerror(errno));
			continue;
		}
		if(grab) {
			if(ioctl(newdev->fd, EVIOCGRAB, 1) < 0) {
				close(newdev->fd);
				free(newdev);
				fprintf(stderr, "Failed to grab %s: %s\n", argv[i], strerror(errno));
				continue;
			}
		}
		newdev->name = basename(strdup(argv[i]));
		newdev->next = evdevs;
		evdevs = newdev;
	}
}
	
static bool add_unixsocket(void) {
	struct sockaddr_un sa = {0};
	sockfd = socket(AF_UNIX, SOCK_STREAM, 0);

	if(sockfd < 0) {
		fprintf(stderr, "Unable to create an AF_UNIX socket: %s\n", strerror(errno));
		return false;
	}

	sa.sun_family = AF_UNIX;
	strncpy(sa.sun_path, device, sizeof sa.sun_path - 1);

	unlink(device);

	if(bind(sockfd, (struct sockaddr *)&sa, sizeof sa) < 0) {
		fprintf(stderr, "Unable to bind AF_UNIX socket to %s: %s\n", device, strerror(errno));
		return false;
	}

	chmod(device, 0666);

	if(listen(sockfd, 3) < 0) {
		fprintf(stderr, "Unable to listen on AF_UNIX socket: %s\n", strerror(errno));
		return false;
	}

	return true;
}


static void processnewclient(void) {
	client_t *newclient = xalloc(sizeof *newclient);

	newclient->fd = accept(sockfd, NULL, NULL);

	if(newclient->fd < 0) {
		free(newclient);
		if(errno == ECONNABORTED || errno == EINTR)
			return;
		syslog(LOG_ERR, "Error during accept(): %s\n", strerror(errno));
		exit(EX_OSERR);
	}

    int flags = fcntl(newclient->fd, F_GETFL);
    fcntl(newclient->fd, F_SETFL, flags | O_NONBLOCK);

	newclient->next = clients;
	clients = newclient;
}

static long time_elapsed(struct timeval *last, struct timeval *current) {
	long seconds = current->tv_sec - last->tv_sec;
	return 1000000 * seconds + current->tv_usec - last->tv_usec;
}

static void processevent(evdev_t *evdev) {
	IRMP_DATA event;
	char hash_key[100];
	char irmp_fulldata[100];
	char message[100];
	int len;
	client_t *client, *prev, *next;

	message[0]=0;
	
	if((len=read(evdev->fd, &event, sizeof event)) <= 0) {
		syslog(LOG_ERR, "Error processing event from %s: %s\n", evdev->name, strerror(errno));
		exit(EX_OSERR);
	}

	DBG ("dummy = 0x%02d, p = %02d, a = 0x%04x, c = 0x%04x, f = 0x%02x\n", event.dummy, event.protocol, event.address, event.command, event.flags);
		
	struct timeval current;
	gettimeofday(&current, NULL);
	if(time_elapsed(&previous_input, &current) < repeat_time)
		repeat++;
	else 
		repeat = 0;

	snprintf (irmp_fulldata, sizeof irmp_fulldata, "%02x%04x%04x%02x", event.protocol, event.address, event.command, event.flags);
	snprintf (hash_key, sizeof hash_key, "%02x%04x%04x%02x", event.protocol, event.address, event.command, 0);

	map_entry_t *map_entry;
	
	if(hashmap_get(mymap, hash_key, (void**)(&map_entry))==MAP_OK) {
		DBG ("MAP_OK irmpd_fulldata=%s lirc=%s\n", irmp_fulldata, map_entry->value);	
		len = snprintf(message, sizeof message, "%s %x %s %s\n",  irmp_fulldata, event.flags, map_entry->value, "IRMP");
	} else {
		DBG ("MAP_ERROR irmpd_fulldata=%s|\n", irmp_fulldata);	
		len = snprintf(message, sizeof message, "%s %x %s %s\n",  irmp_fulldata, repeat, irmp_fulldata, "IRMP");
	}
	
	DBG ("LIRC message=%s\n", message);

	previous_input = current;
	for(client = clients; client; client = client->next) {
		if(write(client->fd, message, len) != len) {
			close(client->fd);
			client->fd = -1;
		}
	}

	for(prev = NULL, client = clients; client; client = next) {
		next = client->next;
		if(client->fd < 0) {
			if(prev)
				prev->next = client->next;
			else
				clients = client->next;
			free(client);
		} else {
			prev = client;
		}
	}
}

static void print_help() {

	printf("irmplircd [-d socket] [-f] [-c] [-r repeat-rate] [-m keycode] -u username] device [device ...]\n\n");
	printf("Options: \n");
	printf("\t-d <socket> UNIX socket. The default is /var/run/lirc/lircd.\n");
	printf("\t-f Run in the foreground.\n");
	printf("\t-r <rate> Repeat rate in ms.\n");
	printf("\t-g Grab the input device(s).\n");
	printf("\t-u <user> User name.\n");
	printf("\t-t <path> Path to translation table.\n");
	printf("\tdevice The input device e.g. /dev/hidraw0\n");
	
}

static void main_loop(void) {
	fd_set permset;
	fd_set fdset;
	evdev_t *evdev;
	int maxfd = 0;

	FD_ZERO(&permset);
	
	for(evdev = evdevs; evdev; evdev = evdev->next) {
		FD_SET(evdev->fd, &permset);
		if(evdev->fd > maxfd)
			maxfd = evdev->fd;
	}
	
	FD_SET(sockfd, &permset);
	if(sockfd > maxfd)
		maxfd = sockfd;

	maxfd++;
	
	while(true) {
		fdset = permset;
		
		if(select(maxfd, &fdset, NULL, NULL, NULL) < 0) {
			if(errno == EINTR)
				continue;
			syslog(LOG_ERR, "Error during select(): %s\n", strerror(errno));
			exit(EX_OSERR);
		}

		for(evdev = evdevs; evdev; evdev = evdev->next)
			if(FD_ISSET(evdev->fd, &fdset))
				processevent(evdev);

		if(FD_ISSET(sockfd, &fdset))
			processnewclient();
	}
}

int main(int argc, char *argv[]) {
	char *user = "nobody";
	char *translation_path = NULL;
	int opt;
	bool foreground = false;
	bool use_translationtable = false;
	
	gettimeofday(&previous_input, NULL);

	while((opt = getopt(argc, argv, "d:gm:fu:r:t:")) != -1) {
        switch(opt) {
			case 'd':
				device = strdup(optarg);
				break;
			case 'g':
				grab = true;
				break;
			case 'u':
				user = strdup(optarg);
				break;
			case 'f':
				foreground = true;
				break;
			case 'r':
				repeat_time = atoi(optarg) * 1000L;
				break;
			case 't':
				use_translationtable = true;
				translation_path = strdup(optarg);
				break;
            default:
				print_help();
                return EX_USAGE;
        }
    }

	if(argc <= optind) {
		fprintf(stderr, "Not enough arguments.\n");
		return EX_USAGE;
	}

	add_evdevs(argc - optind, argv + optind);

	if(!evdevs) {
		fprintf(stderr, "Unable to open any event device!\n");
		return EX_OSERR;
	}

	mymap = hashmap_new();
	
	if(use_translationtable &&  !parse_translation_table(translation_path, mymap)) {
		hashmap_free(mymap);
		return EX_OSERR;
	}

	if (!add_unixsocket()) {
		hashmap_free(mymap);
		if (sockfd >= 0) close (sockfd);
		return EX_OSERR;
	}

	struct passwd *pwd = getpwnam(user);
	if(!pwd) {
		fprintf(stderr, "Unable to resolve user %s!\n", user);
		hashmap_free(mymap);
		if (sockfd >= 0) close (sockfd);
		return EX_OSERR;
	}

	if(setgid(pwd->pw_gid) || setuid(pwd->pw_uid)) {
		fprintf(stderr, "Unable to setuid/setguid to %s!\n", user);
		hashmap_free(mymap);
		if (sockfd >= 0) close (sockfd);
		return EX_OSERR;
	}

	if(!foreground)
		daemon(0, 0);

	syslog(LOG_INFO, "Started");

	signal(SIGPIPE, SIG_IGN);

	main_loop();

	/* Now, destroy the map */
	hashmap_free(mymap);
	if (sockfd >= 0) close (sockfd);

	return 0;
}


irmpexec.c
/*
    irmpexec -- LIRC client that reads IRMP events from the USB IR Remote Receiver 
                and executes a command
	        http://www.mikrocontroller.net/articles/USB_IR_Remote_Receiver
    Copyright (C) 2012  Dirk E. Wagner

    This program is free software; you can redistribute it and/or modify it
    under the terms of version 2 of the GNU General Public License as published
    by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
*/


 /* Standard headers */
#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sysexits.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <syslog.h>
#include <pwd.h>
#include <ctype.h>

/* Input subsystem interface */
#include <linux/input.h>

/* Unix socket support */
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>

#include <sys/poll.h>
#include <sys/file.h>

#include "debug.h"
#include "c_hashmap/hashmap.h"
#include "mapping.h"

static int lirc_fd = -1;
static map_t mymap;
static const char RemoteName[] = "IRMP-exec";

static struct sockaddr_un sa= {
	.sun_family = AF_UNIX,
	.sun_path = "/var/run/lirc/lircd"
};

static bool add_unixsocket(void) {
	lirc_fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (lirc_fd < 0) {
		syslog (LOG_ERR, "Unable to create an AF_UNIX socket: %s", strerror(errno)); 
		return false;
	}
	
	fcntl (lirc_fd, F_SETFL, fcntl (lirc_fd, F_GETFL, 0));

	if (connect (lirc_fd, (struct sockaddr*)&sa, sizeof(sa)) < 0) {
		syslog (LOG_ERR, "Unable to connect AF_UNIX socket %s: %s", sa.sun_path, strerror(errno));
		close(lirc_fd);
		lirc_fd = -1;
		return false;
	}
	
	return true;
}

static int file_ready(int FileDes, int TimeoutMs) {
	fd_set set;
	struct timeval timeout;
	FD_ZERO(&set);
	FD_SET(FileDes, &set);
	if (TimeoutMs >= 0) {
		if (TimeoutMs < 100)
       			TimeoutMs = 100;
		timeout.tv_sec  = TimeoutMs / 1000;
		timeout.tv_usec = (TimeoutMs % 1000) * 1000;
     	}
	return select (FD_SETSIZE, &set, NULL, NULL, (TimeoutMs >= 0) ? &timeout : NULL) > 0 && FD_ISSET(FileDes, &set);
}

static int safe_read(int filedes, void *buffer, int size) {
	for (;;) {
		int p = read(filedes, buffer, size);
		if (p < 0 && errno == EINTR) {
			syslog (LOG_ERR, "EINTR while reading from file handle %d - retrying", filedes);
			continue;
		}
		return p;
	}
}

static int wait_for_keycode (char *lircbuf) {

	int size = -1;
	int ready = 0;
	
	ready = file_ready(lirc_fd, -1);

	size = ready ? safe_read (lirc_fd, lircbuf, BUFSIZ) : -1;
	
//	DBG ("ready=%d size=%d\n", ready, size);
		
	if (ready && size <= 0) {
		syslog (LOG_ERR, "LIRC connection broken. Try to reconnect");
		close(lirc_fd);
		lirc_fd = -1;
		while (lirc_fd < 0) {
			sleep (3);
			if (add_unixsocket()) {
				syslog (LOG_INFO, "Reconnected to LIRC\n");
				break;
			}
		}
	} else 

	if (ready && size > 0) {
		lircbuf [size-1] = '\0';
	}

	return size;
}

static void print_help() {

	printf ("irmpexec [-w] [-d socket] [-f] [-u username] [-t path]\n\n");
	printf ("Options: \n");
	printf ("\t-d <socket> UNIX socket. The default is /var/run/lirc/lircd.\n");
	printf ("\t-f Run in the foreground.\n");
	printf ("\t-u <user> User name.\n");
	printf ("\t-t <path> Path to translation table.\n");
	printf ("\t-w lirc irw like mode, print data.\n");
	
}

static void main_loop(bool irw_mode) {

	int repeat = 0;
	int bufSize = 0;
	char lircbuf[BUFSIZ] = "";
	char irmp_data[BUFSIZ] = "";
	char irmp_code[BUFSIZ] = "";
	char id[BUFSIZ] = ""; 
	
	if (irw_mode) {
		printf ("irmpdata\t|repeat\t|irmp_codes\t|id\n"); 
	}

	do {
		memset(&lircbuf, 0, sizeof(lircbuf));
		if ((bufSize = wait_for_keycode (lircbuf)) > 0) {
			sscanf (lircbuf, "%s %d %s %s", irmp_data, &repeat, irmp_code, id);
			DBG ("irmpdata=%s repeat=%d irmp_code=%s id=%s\n", irmp_data, repeat, irmp_code, id);
			if (strncasecmp (id, "IRMP", 4) == 0) {
				if (irw_mode) {
					printf ("%s\t|%d\t|%s\t|%s\n", irmp_data, repeat, irmp_code, id);
				} else {
					if (!repeat) {

						map_entry_t *map_entry;
			
						if(hashmap_get(mymap, irmp_code, (void**)(&map_entry))==MAP_OK) {
							DBG ("MAP_OK map_entry->irmp_code=%s map_entry->value=%s\n", map_entry->key, map_entry->value);	
							syslog(LOG_INFO, "executing by IRMP (%s)", map_entry->value);
							system (map_entry->value);
						} else {
							DBG ("MAP_ERROR irmp_code=%s|\n", irmp_code);	
						}

					} else {
						DBG ("Repetition ignored\n");
					}
				}
			} else {
				DBG ("Wrong ID %s, code ignored\n", id);
			}
			
		}
	} while (true);

}

int main(int argc, char *argv[]) {
	
	char *user = "nobody";
	char *translation_path = "/etc/irmpexec.map";
	int opt;
	bool foreground = false;
	bool irw_mode = false;
	bool use_translationtable = false;
	
	while ((opt = getopt(argc, argv, "whd:fu:t:")) != -1) {
        	switch (opt) {
			case 'd':
				strncpy (sa.sun_path, optarg, sizeof sa.sun_path - 1); 
				break;
			case 'u':
				user = strdup (optarg);
				break;
			case 'f':
				foreground = true;
				break;
			case 't':
				use_translationtable = true;
				translation_path = strdup (optarg);
				break;
			case 'w':
				irw_mode = true;
				foreground = true;
				break;
			case 'h':
				print_help ();
				return 0;
            		default:
				print_help ();
                		return EX_USAGE;
        	}
    	}

	mymap = hashmap_new();
	
	if (use_translationtable && !parse_translation_table(translation_path, mymap)) {
		hashmap_free(mymap);
		return EX_OSERR;
	}

	if (!add_unixsocket()) {
		hashmap_free(mymap);
		if (lirc_fd >= 0) close(lirc_fd);
		return EX_OSERR;
	}

	struct passwd *pwd = getpwnam(user);
	if(!pwd) {
		hashmap_free(mymap);
		if (lirc_fd >= 0) close(lirc_fd);
		fprintf(stderr, "Unable to resolve user %s!\n", user);
		return EX_OSERR;
	}

	if(setgid(pwd->pw_gid) || setuid(pwd->pw_uid)) {
		hashmap_free(mymap);
		if (lirc_fd >= 0) close(lirc_fd);
		fprintf(stderr, "Unable to setuid/setguid to %s!\n", user);
		return EX_OSERR;
	}

	if(!foreground)
		daemon(0, 0);

	syslog(LOG_INFO, "Started");

	signal(SIGPIPE, SIG_IGN);

	main_loop(irw_mode);

	/* Now, destroy the map */
	hashmap_free(mymap);
	if (lirc_fd >= 0) close(lirc_fd);
	
	return 0;
}
